rosbag2_transport/src/rosbag2_transport/recorder_event_notifier_impl.hpp

```diff
@@ -0,0 +1,236 @@
+// Copyright 2025 Apex.AI, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+#ifndef ROSBAG2_TRANSPORT__RECORDER_EVENT_NOTIFIER_IMPL_HPP_
+#define ROSBAG2_TRANSPORT__RECORDER_EVENT_NOTIFIER_IMPL_HPP_
+
+#include <chrono>
+#include <condition_variable>
+#include <mutex>
+#include <queue>
+#include <stdexcept>
+#include <string>
+#include <thread>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "rclcpp/logging.hpp"
+#include "rclcpp/node.hpp"
+
+#include "rosbag2_interfaces/msg/write_split_event.hpp"
+#include "rosbag2_cpp/bag_events.hpp"
+#include "rosbag2_transport/recorder_event_notifier.hpp"
+
+namespace rosbag2_transport
+{
+class RecorderEventNotifierImpl
+{
+public:
+  explicit RecorderEventNotifierImpl(rclcpp::Node * node)
+  : node(node)
+  {
+    if (!node) {
+      throw std::invalid_argument("Node pointer cannot be null");
+    }
+    split_event_pub_ =
+      node->create_publisher<rosbag2_interfaces::msg::WriteSplitEvent>("events/write_split", 1);
+
+    // Start the thread that will publish events
+    {
+      std::lock_guard<std::mutex> lock(event_publisher_thread_mutex_);
+      event_publisher_thread_should_exit_ = false;
+      event_publisher_thread_ = std::thread(&RecorderEventNotifierImpl::event_publisher_thread_main,
+                                            this);
+    }
+  }
+
+  virtual ~RecorderEventNotifierImpl()
+  {
+    if (event_publisher_thread_.joinable()) {
+      {
+        std::lock_guard<std::mutex> lock(event_publisher_thread_mutex_);
+        event_publisher_thread_should_exit_ = true;
+      }
+      event_publisher_thread_wake_cv_.notify_all();
+      event_publisher_thread_.join();
+    }
+  }
+
+  void set_messages_lost_statistics_max_publishing_rate(float update_rate_hz)
+  {
+    {
+      std::unique_lock<std::mutex> pub_thread_lock(event_publisher_thread_mutex_);
+      if (update_rate_hz == 0.0f) {
+        disable_publishing_msgs_lost_statistics_ = true;
+        RCLCPP_DEBUG(node->get_logger(), "Messages lost statistics publishing is disabled");
+      } else if (update_rate_hz > 0.0f) {
+        disable_publishing_msgs_lost_statistics_ = false;
+        msgs_lost_stats_max_publishing_period_ =
+          std::chrono::milliseconds(static_cast<int>(1000 / update_rate_hz));
```

Comment:
@MichaelOrlov is there any validation of `update_rate_hz`? If it is set (probably mistakenly) to more than 1000, the period will be 0 and I think you will busy loop.

Reply:
@kylemarcey Good catch I will add an extra check and throw an exception